# EDA Smartphones

### Librerías

In [1]:
import pandas as pd
from utils import funciones as f

# pd.set_option('display.max_colwidth', None)

### Data

In [2]:
df = pd.read_csv('.\data\smartphonesmessy.csv')

In [3]:
df.shape

(1020, 11)

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1020 entries, 0 to 1019
Data columns (total 11 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   model      1020 non-null   object 
 1   price      1020 non-null   object 
 2   rating     879 non-null    float64
 3   sim        1020 non-null   object 
 4   processor  1020 non-null   object 
 5   ram        1020 non-null   object 
 6   battery    1020 non-null   object 
 7   display    1020 non-null   object 
 8   camera     1019 non-null   object 
 9   card       1013 non-null   object 
 10  os         1003 non-null   object 
dtypes: float64(1), object(10)
memory usage: 87.8+ KB


In [5]:
df.describe(include = 'all').round()

model    price  rating                             sim  \
count            1020     1020   879.0                            1020   
unique           1020      412     NaN                              28   
top     OnePlus 11 5G  ₹14,999     NaN  Dual Sim, 3G, 4G, VoLTE, Wi-Fi   
freq                1       21     NaN                             324   
mean              NaN      NaN    78.0                             NaN   
std               NaN      NaN     7.0                             NaN   
min               NaN      NaN    60.0                             NaN   
25%               NaN      NaN    74.0                             NaN   
50%               NaN      NaN    80.0                             NaN   
75%               NaN      NaN    84.0                             NaN   
max               NaN      NaN    89.0                             NaN   

                                             processor  \
count                                             1020   
unique                                             298   
top     Dimensity 700 5G, Octa Core, 2.2 GHz Processor   
freq                                                29   
mean                                               NaN   
std                                                NaN   
min                                                NaN   
25%                                                NaN   
50%                                                NaN   
75%                                                NaN   
max                                                NaN   

                             ram                                  battery  \
count                       1020                                     1020   
unique                        58                                      256   
top     8 GB RAM, 128 GB inbuilt  5000 mAh Battery with 33W Fast Charging   
freq                         267                                      103   
mean                         NaN                                      NaN   
std                          NaN                                      NaN   
min                          NaN                                      NaN   
25%                          NaN                                      NaN   
50%                          NaN                                      NaN   
75%                          NaN                                      NaN   
max                          NaN                                      NaN   

                                                  display  \
count                                                1020   
unique                                                369   
top     6.67 inches, 1080 x 2400 px, 120 Hz Display wi...   
freq                                                   54   
mean                                                  NaN   
std                                                   NaN   
min                                                   NaN   
25%                                                   NaN   
50%                                                   NaN   
75%                                                   NaN   
max                                                   NaN   

                                                   camera  \
count                                                1019   
unique                                                285   
top     50 MP + 8 MP + 2 MP Triple Rear & 16 MP Front ...   
freq                                                   40   
mean                                                  NaN   
std                                                   NaN   
min                                                   NaN   
25%                                                   NaN   
50%                                                   NaN   
75%                                                   NaN   
max                                                   NaN   

                                    car

In [6]:
f.cardinalidad(df)

variable tipo_dato  cantidad_de_nulos  \
0       model    object                  0   
1       price    object                  0   
2      rating   float64                141   
3         sim    object                  0   
4   processor    object                  0   
5         ram    object                  0   
6     battery    object                  0   
7     display    object                  0   
8      camera    object                  1   
9        card    object                  7   
10         os    object                 17   

                                       valores_unicos  cardinalidad  \
0   [OnePlus 11 5G, OnePlus Nord CE 2 Lite 5G, Sam...          1020   
1   [₹54,999, ₹19,989, ₹16,499, ₹14,999, ₹24,999, ...           412   
2   [89.0, 81.0, 75.0, 82.0, 80.0, 86.0, 85.0, 84....            30   
3   [Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, Dual...            28   
4   [Snapdragon 8 Gen2, Octa Core, 3.2 GHz Process...           298   
5   [12 GB RAM, 256 GB inbuilt, 6 GB RAM, 128 GB i...            58   
6   [5000 mAh Battery with 100W Fast Charging, 500...           256   
7   [6.7 inches, 1440 x 3216 px, 120 Hz Display wi...           369   
8   [50 MP + 48 MP + 32 MP Triple Rear & 16 MP Fro...           285   
9   [Memory Card Not Supported, Memory Card (Hybri...            63   
10  [Android v13, Android v12, iOS v16, No FM Radi...            48   

    porcentaje_cardinalidad tipo_variable  
0                100.000000       nominal  
1                 40.392157      continua  
2                  2.941176      discreta  
3                  2.745098       nominal  
4                 29.215686       nominal  
5                  5.686275       ordinal  
6                 25.098039       ordinal  
7                 36.176471       nominal  
8                 27.941176       ordinal  
9                  6.176471       ordinal  
10                 4.705882       nominal

##### Al analizar los valores dentro de ciertas columnas podemos concluir:
- La columna 'ram' tiene datos que no corresponden a por ejemplo: 'Dual Core, 500\u2009MHz Processor'
- La columna 'processor' con tiene datos que no se corresponden los demás por ejemplo: '4\u2009MB RAM, 4\u2009MB inbuilt', '1450\u2009mAh Battery'
- La columna 'ram' tiene datos que no se corresponden con lo demás, por ejemplo: '1.5 inches, 120\u2009x\u2009120\u2009px Display'
- La columna 'battery' tiene datos que no se corresponden con los demás, por ejemplo: '48\u2009MB RAM, 128\u2009MB inbuilt'
- La columna 'display' tiene datos que no se corresponden con los demás, por ejemplo: '0.3\u2009MP Rear Camera'
- La columna 'camera' tiene datos que no se corresponden con los demás, por ejemplo: '1.77 inches, 240\u2009x\u2009320\u2009px Display'
- La columna 'card' tiene datos que no se corresponden con los demás, por ejemplo: '50\u2009MP + 48\u2009MP + 32\u2009MP Triple Rear & 32\u2009MP + 32\u2009MP Dual Front Camera'
- La columna 'os' tiene datos que no se corresponden con los demás, por ejemplo: 'Memory Card Supported, upto 32\u2009GB'
- La columna 'price' esta en rupias

In [20]:
df[df['rating'].isna()]

model      price  rating  \
14    Samsung Galaxy S23 Ultra 5G  ₹1,14,990     NaN   
29                 OnePlus 11 Pro    ₹69,999     NaN   
37    Samsung Galaxy S22 Ultra 5G    ₹91,999     NaN   
49          Samsung Galaxy A74 5G    ₹42,999     NaN   
69               Oppo Find N Fold    ₹99,990     NaN   
...                           ...        ...     ...   
996                     Vivo Y55S    ₹13,490     NaN   
1000              XTouch F40 Flip     ₹1,999     NaN   
1003                      Lava X3     ₹6,999     NaN   
1012                    itel A23s     ₹4,787     NaN   
1014                Vivo X Fold 2  ₹1,19,990     NaN   

                                                sim  \
14    Dual Sim, 3G, 4G, 5G, VoLTE, Vo5G, Wi-Fi, NFC   
29          Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC   
37          Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC   
49          Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC   
69          Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC   
...                                             ...   
996                  Dual Sim, 3G, 4G, VoLTE, Wi-Fi   
1000                                       Dual Sim   
1003                 Dual Sim, 3G, 4G, VoLTE, Wi-Fi   
1012                        Dual Sim, 3G, 4G, Wi-Fi   
1014        Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC   

                                              processor  \
14      Snapdragon 8 Gen2, Octa Core, 3.2 GHz Processor   
29     Snapdragon  8 Gen2, Octa Core, 3.2 GHz Processor   
37       Snapdragon  8 Gen1, Octa Core, 3 GHz Processor   
49     Snapdragon 7 Gen1, Octa Core, 2.36 GHz Processor   
69    Snapdragon  8+ Gen1, Octa Core, 3.2 GHz Processor   
...                                                 ...   
996        Snapdragon 425, Quad Core, 1.2 GHz Processor   
1000                                              No 3G   
1003              Helio A22, Quad Core, 2 GHz Processor   
1012  Spreadtrum  SC9832E, Quad Core, 1.4 GHz Processor   
1014    Snapdragon 8 Gen2, Octa Core, 3.2 GHz Processor   

                            ram                                   battery  \
14     8 GB RAM, 256 GB inbuilt   5000 mAh Battery with 45W Fast Charging   
29     8 GB RAM, 128 GB inbuilt  4500 mAh Battery with 100W Fast Charging   
37    12 GB RAM, 256 GB inbuilt   5000 mAh Battery with 45W Fast Charging   
49     8 GB RAM, 128 GB inbuilt   5000 mAh Battery with 33W Fast Charging   
69     8 GB RAM, 256 GB inbuilt   5000 mAh Battery with 67W Fast Charging   
...                         ...                                       ...   
996     3 GB RAM, 16 GB inbuilt                          2730 mAh Battery   
1000                    No Wifi                  32 MB RAM, 32 MB inbuilt   
1003    3 GB RAM, 32 GB inbuilt                          4000 mAh Battery   
1012    2 GB RAM, 32 GB inbuilt                          3020 mAh Battery   
1014  12 GB RAM, 256 GB inbuilt   4800 mAh Battery with 66W Fast Charging   

                                                display  \
14    6.8 inches, 1440 x 3088 px, 120 Hz Display wit...   
29    6.7 inches, 1440 x 3216 px, 120 Hz Display wit...   
37    6.8 inches, 1440 x 3080 px, 120 Hz Display wit...   
49    6.7 inches, 1080 x 2412 px, 120 Hz Display wit...   
69    7.1 inches, 1792 x 1920 px, 120 Hz Display wit...   
...                                                 ...   
996                   5.2 inches, 720 x 1280 px Display   
1000                                    800 mAh Battery   
1003  6.5 inches, 1600 x 720 px Display with Water D...   
1012                     5 inches, 854 x 480 px Display   
1014        8.03 inches, 1916 x 2160 px, 120 Hz Display   

                                                 camera  \
14                200 MP Quad Rear & 12 MP Front Camera   
29    50 MP + 48 MP + 32 MP Triple Rear & 16 MP Fron...   
37                108 MP Quad Rear & 40 MP Front Camera   
49                108 MP Quad Rear & 32 MP Front Camera   
69                       Foldable Display, Dual Display 

In [21]:
df[df['model'] == 'XTouch F40 Flip']

model   price  rating       sim processor      ram  \
1000  XTouch F40 Flip  ₹1,999     NaN  Dual Sim     No 3G  No Wifi   

                       battery          display  \
1000  32 MB RAM, 32 MB inbuilt  800 mAh Battery   

                                 camera          card                  os  
1000  1.77 inches, 240 x 320 px Display  Dual Display  1.3 MP Rear Camera

In [24]:
df[~df['processor'].str.contains('Processor')]

model      price  rating  \
152         Apple iPhone 15 Pro Max  ₹1,42,990    79.0   
263                Apple iPhone XR2    ₹71,999    60.0   
282            Apple iPhone 14 Mini    ₹69,990    70.0   
376                 Nokia 2660 Flip     ₹4,649     NaN   
429                    Nokia X50 5G    ₹34,999    76.0   
448             Apple iPhone 15 Pro  ₹1,30,990    75.0   
473                    Nokia 110 4G     ₹1,762     NaN   
532   Samsung Guru Music 2 Dual Sim     ₹1,949     NaN   
551                 itel Magic 2 4G     ₹2,399     NaN   
573                Nokia 105 (2019)     ₹1,299     NaN   
582                   Nokia 8210 4G     ₹3,749     NaN   
608               Namotel Achhe Din        ₹99     NaN   
611                 Eunity U1 5605N       ₹699     NaN   
627                    Nokia 225 4G     ₹3,589     NaN   
640                  Nokia 105 Plus     ₹1,299     NaN   
645                 Nokia 2760 Flip     ₹5,490     NaN   
647               Motorola Moto A10     ₹1,339     NaN   
657                   Zanco Tiny T1     ₹2,799     NaN   
665                    itel it2163S       ₹958     NaN   
685                    Xiaomi Qin 1     ₹1,799     NaN   
699           Samsung Guru GT-E1215     ₹1,850     NaN   
728                 Apple iPhone 15    ₹82,990    72.0   
748                    Nokia 400 4G     ₹3,290     NaN   
754      Apple iPod Touch (7th Gen)    ₹18,900     NaN   
757                    Karbonn KU3i       ₹995     NaN   
817                    itel Magic X     ₹2,239     NaN   
852                    Nokia 215 4G     ₹3,099     NaN   
857                     LG Folder 2    ₹11,999     NaN   
881                Nokia 110 (2022)     ₹1,599     NaN   
882          Nokia 5710 XpressAudio     ₹4,799     NaN   
917                Nokia 150 (2020)     ₹2,499     NaN   
925                   Nokia 3310 4G     ₹3,999     NaN   
927              Samsung Guru E1200     ₹1,685     NaN   
930                iKall K88 Pro 4G     ₹2,099     NaN   
988             Nokia 5310 Dual Sim     ₹3,399     NaN   
1000                XTouch F40 Flip     ₹1,999     NaN   

                                          sim                   processor  \
152   Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC                 Bionic  A16   
263       Dual Sim, 3G, 4G, VoLTE, Wi-Fi, NFC                  Apple  A13   
282       Dual Sim, 3G, 4G, VoLTE, Wi-Fi, NFC                  Bionic A16   
376                          Dual Sim, 3G, 4G                     No Wifi   
429      Single Sim, 3G, 4G, 5G, VoLTE, Wi-Fi              Snapdragon 775   
448   Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC                 Bionic  A16   
473                   Dual Sim, 3G, 4G, VoLTE                     No Wifi   
532                                  Dual Sim                     No Wifi   
551            Dual Sim, 3G, 4G, VoLTE, Wi-Fi                       No 3G   
573                                Single Sim                     No Wifi   
582                          Dual Sim, 3G, 4G                     No Wifi   
608                       Dual Sim, 3G, Wi-Fi      1 GB RAM, 4 GB inbuilt   
611                                  Dual Sim                       No 3G   
627                   Dual Sim, 3G, 4G, VoLTE                 Unisoc T117   
640                                  Dual Sim      4 MB RAM, 4 MB inbuilt   
645                   Dual Sim, 3G, 4G, Wi-Fi            1450 mAh Battery   
647                                  Dual Sim      4 MB RAM, 4 MB inbuilt   
657                                Single Sim    32 MB RAM, 32 MB inbuilt   
665                                  Dual Sim      4 MB RAM, 4 MB inbuilt   
685                           Dual Sim, Wi-Fi               Helio MT6260A   
699                                Single Sim             800 mAh Battery   
728       Dual Sim, 3G, 4G, VoLTE, Wi-Fi, NFC                 Bionic  A16   
748                Dual Sim, 4G, VoLTE, Wi-Fi            2000 mAh Battery   
754                                

In [25]:
df['card']

0              Memory Card Not Supported
1        Memory Card (Hybrid), upto 1 TB
2       Memory Card Supported, upto 1 TB
3        Memory Card (Hybrid), upto 1 TB
4              Memory Card Not Supported
                      ...               
1015                         Android v12
1016    Memory Card Supported, upto 1 TB
1017           Memory Card Not Supported
1018    Memory Card Supported, upto 1 TB
1019    Memory Card Supported, upto 1 TB
Name: card, Length: 1020, dtype: object

In [26]:
df[df['rating'].isna()]

model      price  rating  \
14    Samsung Galaxy S23 Ultra 5G  ₹1,14,990     NaN   
29                 OnePlus 11 Pro    ₹69,999     NaN   
37    Samsung Galaxy S22 Ultra 5G    ₹91,999     NaN   
49          Samsung Galaxy A74 5G    ₹42,999     NaN   
69               Oppo Find N Fold    ₹99,990     NaN   
...                           ...        ...     ...   
996                     Vivo Y55S    ₹13,490     NaN   
1000              XTouch F40 Flip     ₹1,999     NaN   
1003                      Lava X3     ₹6,999     NaN   
1012                    itel A23s     ₹4,787     NaN   
1014                Vivo X Fold 2  ₹1,19,990     NaN   

                                                sim  \
14    Dual Sim, 3G, 4G, 5G, VoLTE, Vo5G, Wi-Fi, NFC   
29          Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC   
37          Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC   
49          Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC   
69          Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC   
...                                             ...   
996                  Dual Sim, 3G, 4G, VoLTE, Wi-Fi   
1000                                       Dual Sim   
1003                 Dual Sim, 3G, 4G, VoLTE, Wi-Fi   
1012                        Dual Sim, 3G, 4G, Wi-Fi   
1014        Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC   

                                              processor  \
14      Snapdragon 8 Gen2, Octa Core, 3.2 GHz Processor   
29     Snapdragon  8 Gen2, Octa Core, 3.2 GHz Processor   
37       Snapdragon  8 Gen1, Octa Core, 3 GHz Processor   
49     Snapdragon 7 Gen1, Octa Core, 2.36 GHz Processor   
69    Snapdragon  8+ Gen1, Octa Core, 3.2 GHz Processor   
...                                                 ...   
996        Snapdragon 425, Quad Core, 1.2 GHz Processor   
1000                                              No 3G   
1003              Helio A22, Quad Core, 2 GHz Processor   
1012  Spreadtrum  SC9832E, Quad Core, 1.4 GHz Processor   
1014    Snapdragon 8 Gen2, Octa Core, 3.2 GHz Processor   

                            ram                                   battery  \
14     8 GB RAM, 256 GB inbuilt   5000 mAh Battery with 45W Fast Charging   
29     8 GB RAM, 128 GB inbuilt  4500 mAh Battery with 100W Fast Charging   
37    12 GB RAM, 256 GB inbuilt   5000 mAh Battery with 45W Fast Charging   
49     8 GB RAM, 128 GB inbuilt   5000 mAh Battery with 33W Fast Charging   
69     8 GB RAM, 256 GB inbuilt   5000 mAh Battery with 67W Fast Charging   
...                         ...                                       ...   
996     3 GB RAM, 16 GB inbuilt                          2730 mAh Battery   
1000                    No Wifi                  32 MB RAM, 32 MB inbuilt   
1003    3 GB RAM, 32 GB inbuilt                          4000 mAh Battery   
1012    2 GB RAM, 32 GB inbuilt                          3020 mAh Battery   
1014  12 GB RAM, 256 GB inbuilt   4800 mAh Battery with 66W Fast Charging   

                                                display  \
14    6.8 inches, 1440 x 3088 px, 120 Hz Display wit...   
29    6.7 inches, 1440 x 3216 px, 120 Hz Display wit...   
37    6.8 inches, 1440 x 3080 px, 120 Hz Display wit...   
49    6.7 inches, 1080 x 2412 px, 120 Hz Display wit...   
69    7.1 inches, 1792 x 1920 px, 120 Hz Display wit...   
...                                                 ...   
996                   5.2 inches, 720 x 1280 px Display   
1000                                    800 mAh Battery   
1003  6.5 inches, 1600 x 720 px Display with Water D...   
1012                     5 inches, 854 x 480 px Display   
1014        8.03 inches, 1916 x 2160 px, 120 Hz Display   

                                                 camera  \
14                200 MP Quad Rear & 12 MP Front Camera   
29    50 MP + 48 MP + 32 MP Triple Rear & 16 MP Fron...   
37                108 MP Quad Rear & 40 MP Front Camera   
49                108 MP Quad Rear & 32 MP Front Camera   
69                       Foldable Display, Dual Display 

In [27]:
df[df['sim'] == 'Dual Sim']

model   price  rating       sim  \
532   Samsung Guru Music 2 Dual Sim  ₹1,949     NaN  Dual Sim   
611                 Eunity U1 5605N    ₹699     NaN  Dual Sim   
640                  Nokia 105 Plus  ₹1,299     NaN  Dual Sim   
647               Motorola Moto A10  ₹1,339     NaN  Dual Sim   
665                    itel it2163S    ₹958     NaN  Dual Sim   
757                    Karbonn KU3i    ₹995     NaN  Dual Sim   
835                  Eunity U1 King    ₹599     NaN  Dual Sim   
846                   DIZO Star 500  ₹1,749     NaN  Dual Sim   
881                Nokia 110 (2022)  ₹1,599     NaN  Dual Sim   
917                Nokia 150 (2020)  ₹2,499     NaN  Dual Sim   
932                  Eunity U1 Livo    ₹649     NaN  Dual Sim   
988             Nokia 5310 Dual Sim  ₹3,399     NaN  Dual Sim   
1000                XTouch F40 Flip  ₹1,999     NaN  Dual Sim   

                      processor                             ram  \
532                     No Wifi  Single Core, 208 MHz Processor   
611                       No 3G    Dual Core, 500 MHz Processor   
640      4 MB RAM, 4 MB inbuilt                 800 mAh Battery   
647      4 MB RAM, 4 MB inbuilt                1750 mAh Battery   
665      4 MB RAM, 4 MB inbuilt                1200 mAh Battery   
757    52 MB RAM, 32 MB inbuilt                1000 mAh Battery   
835          1.77 MHz Processor        32 MB RAM, 32 MB inbuilt   
846   SC6531E, 26 MHz Processor        32 MB RAM, 32 MB inbuilt   
881                     No Wifi          4 MB RAM, 4 MB inbuilt   
917                     No Wifi                    4 MB inbuilt   
932          1.77 MHz Processor        16 MB RAM, 16 MB inbuilt   
988     8 MB RAM, 16 MB inbuilt                1200 mAh Battery   
1000                      No 3G                         No Wifi   

                                battery                            display  \
532                     800 mAh Battery     2 inches, 128 x 160 px Display   
611            32 MB RAM, 32 MB inbuilt                   3000 mAh Battery   
640   1.77 inches, 128 x 160 px Display                     No Rear Camera   
647    1.8 inches, 160 x 128 px Display                     No Rear Camera   
665    1.8 inches, 160 x 128 px Display                     No Rear Camera   
757    1.8 inches, 128 x 160 px Display                     No Rear Camera   
835                    1000 mAh Battery   1.8 inches, 260 x 240 px Display   
846                    1900 mAh Battery   2.8 inches, 240 x 320 px Display   
881                    1000 mAh Battery  1.77 inches, 120 x 160 px Display   
917                    1020 mAh Battery   2.4 inches, 240 x 320 px Display   
932                    1000 mAh Battery   1.8 inches, 280 x 240 px Display   
988    2.4 inches, 240 x 320 px Display                 0.3 MP Rear Camera   
1000           32 MB RAM, 32 MB inbuilt                    800 mAh Battery   

                                 camera                               card  \
532                      No Rear Camera  Memory Card Supported, upto 16 GB   
611    1.8 inches, 128 x 160 px Display                 0.3 MP Rear Camera   
640   Memory Card Supported, upto 32 GB                          Bluetooth   
647   Memory Card Supported, upto 32 GB                                NaN   
665   Memory Card Supported, upto 32 GB                          Bluetooth   
757   Memory Card Supported, upto 16 GB                          Bluetooth   
835                  0.3 MP Rear Camera                                NaN   
846                  0.3 MP Rear Camera  Memory Card Supported, upto 64 GB   
881   Memory Card Supported, upto 32 GB                                NaN   
917                  0.3 MP Rear Camera  Memory Card Supported, upto 32 GB   
932                  0.3 MP Rear Camera                                NaN   
988   Memory Card Supported, upto 32 GB                          Bluetooth   
1000  1.77 inches, 240 x 320 px Display                       Dual Displa

In [29]:
df[df['sim'] == 'Wi-Fi']

model    price  rating    sim      processor  \
754  Apple iPod Touch (7th Gen)  ₹18,900     NaN  Wi-Fi  32 GB inbuilt   

                                 ram                          battery  \
754  4 inches, 640 x 1136 px Display  8 MP Rear & 1.2 MP Front Camera   

     display       camera       card       os  
754  iOS v12  No FM Radio  Bluetooth  Browser

In [28]:
df['sim'].unique()

array(['Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC',
       'Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi',
       'Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, IR Blaster',
       'Dual Sim, 3G, 4G, 5G, VoLTE, Vo5G, Wi-Fi, NFC',
       'Dual Sim, 3G, 4G, VoLTE, Wi-Fi',
       'Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Blaster',
       'Dual Sim, 3G, 4G, VoLTE, Wi-Fi, NFC',
       'Dual Sim, 3G, 4G, VoLTE, Wi-Fi, IR Blaster',
       'Dual Sim, 3G, 4G, Wi-Fi', 'Single Sim, 3G, 4G, VoLTE, Wi-Fi, NFC',
       'Single Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC',
       'Dual Sim, 3G, 4G, 5G, VoLTE, Vo5G, Wi-Fi', 'Dual Sim, 3G, 4G',
       'Single Sim, 3G, 4G, 5G, VoLTE, Wi-Fi',
       'Single Sim, 3G, 4G, Wi-Fi, NFC', 'Dual Sim, 3G, 4G, VoLTE',
       'Dual Sim, 3G, 4G, VoLTE, Wi-Fi, NFC, IR Blaster', 'Dual Sim',
       'Single Sim', 'Single Sim, 3G, 4G, VoLTE, Wi-Fi',
       'Dual Sim, 3G, Wi-Fi', 'Dual Sim, Wi-Fi',
       'Dual Sim, 3G, 4G, Wi-Fi, NFC', 'Dual Sim, 4G, VoLTE, Wi-Fi',
       'Wi-Fi', 'Dual Sim, 3G, 4

In [30]:
for i in df['sim'].unique():
    print (i)

Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC
Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi
Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, IR Blaster
Dual Sim, 3G, 4G, 5G, VoLTE, Vo5G, Wi-Fi, NFC
Dual Sim, 3G, 4G, VoLTE, Wi-Fi
Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC, IR Blaster
Dual Sim, 3G, 4G, VoLTE, Wi-Fi, NFC
Dual Sim, 3G, 4G, VoLTE, Wi-Fi, IR Blaster
Dual Sim, 3G, 4G, Wi-Fi
Single Sim, 3G, 4G, VoLTE, Wi-Fi, NFC
Single Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC
Dual Sim, 3G, 4G, 5G, VoLTE, Vo5G, Wi-Fi
Dual Sim, 3G, 4G
Single Sim, 3G, 4G, 5G, VoLTE, Wi-Fi
Single Sim, 3G, 4G, Wi-Fi, NFC
Dual Sim, 3G, 4G, VoLTE
Dual Sim, 3G, 4G, VoLTE, Wi-Fi, NFC, IR Blaster
Dual Sim
Single Sim
Single Sim, 3G, 4G, VoLTE, Wi-Fi
Dual Sim, 3G, Wi-Fi
Dual Sim, Wi-Fi
Dual Sim, 3G, 4G, Wi-Fi, NFC
Dual Sim, 4G, VoLTE, Wi-Fi
Wi-Fi
Dual Sim, 3G, 4G, 5G, VoLTE, Vo5G, Wi-Fi, IR Blaster
Single Sim, 3G, 4G, Wi-Fi
Dual Sim, 3G, 4G, 5G, VoLTE, Vo5G, Wi-Fi, NFC, IR Blaster


In [31]:
df[df['sim'] == 'Single Sim']

model   price  rating         sim  \
573       Nokia 105 (2019)  ₹1,299     NaN  Single Sim   
657          Zanco Tiny T1  ₹2,799     NaN  Single Sim   
699  Samsung Guru GT-E1215  ₹1,850     NaN  Single Sim   
927     Samsung Guru E1200  ₹1,685     NaN  Single Sim   

                    processor                               ram  \
573                   No Wifi            4 MB RAM, 4 MB inbuilt   
657  32 MB RAM, 32 MB inbuilt                   200 mAh Battery   
699           800 mAh Battery  1.5 inches, 120 x 120 px Display   
927                   No Wifi                   800 mAh Battery   

                              battery                            display  \
573                   800 mAh Battery  1.77 inches, 120 x 160 px Display   
657   0.49 inches, 64 x 32 px Display                     No Rear Camera   
699                    No Rear Camera                        No FM Radio   
927  1.5 inches, 128 x 128 px Display                     No Rear Camera   

             camera       card   os  
573  No Rear Camera        NaN  NaN  
657     No FM Radio  Bluetooth  NaN  
699             NaN        NaN  NaN  
927     No FM Radio        NaN  NaN

In [37]:
df['camera'].unique()

array(['50\u2009MP + 48\u2009MP + 32\u2009MP Triple Rear & 16\u2009MP Front Camera',
       '64\u2009MP + 2\u2009MP + 2\u2009MP Triple Rear & 16\u2009MP Front Camera',
       '50\u2009MP + 2\u2009MP + 2\u2009MP Triple Rear & 13\u2009MP Front Camera',
       '50\u2009MP + 8\u2009MP + 2\u2009MP Triple Rear & 16\u2009MP Front Camera',
       '108\u2009MP + 8\u2009MP + 2\u2009MP Triple Rear & 16\u2009MP Front Camera',
       '50\u2009MP + 8\u2009MP + 2\u2009MP Triple Rear & 8\u2009MP Front Camera',
       '12\u2009MP + 12\u2009MP Dual Rear & 12\u2009MP Front Camera',
       '200\u2009MP + 8\u2009MP + 2\u2009MP Triple Rear & 16\u2009MP Front Camera',
       '50\u2009MP + 50\u2009MP Dual Rear & 16\u2009MP Front Camera',
       '50\u2009MP + 8\u2009MP + 2\u2009MP Triple Rear & 32\u2009MP Front Camera',
       '108\u2009MP + 2\u2009MP Dual Rear & 16\u2009MP Front Camera',
       '50\u2009MP + 2\u2009MP Dual Rear & 8\u2009MP Front Camera',
       '50\u2009MP + 2\u2009MP + 2\u2009MP Triple Rear 

In [38]:
df[df['camera'].str.contains('Camera')]

ValueError: Cannot mask with non-boolean array containing NA / NaN values